File: src/main/java/net/dean/jraw/models/PublicContribution.java
package net.dean.jraw.models;

import com.fasterxml.jackson.databind.JsonNode;
import net.dean.jraw.models.attr.Distinguishable;
import net.dean.jraw.models.attr.Gildable;
import net.dean.jraw.models.attr.Votable;
import net.dean.jraw.models.meta.ContributionSerializer;
import net.dean.jraw.models.meta.JsonProperty;
import net.dean.jraw.models.meta.Model;

import java.util.HashMap;
import java.util.Map;

/**
 * This class is used to separate public contributions (submissions and comments) from private contributions (messages)
 */
@Model(kind = Model.Kind.ABSTRACT, serializer = ContributionSerializer.class)
public abstract class PublicContribution extends Contribution implements Distinguishable, Gildable, Votable {
    /** Instantiates a new PublicContribution */
    public PublicContribution(JsonNode dataNode) {
        super(dataNode);
    }

    /** Gets a map of reasons to the amount of times reported for that reason by normal users (non-moderators) */
    @JsonProperty(nullable = true)
    public Map<String, Integer> getUserReports() {
        if (!data.has("user_reports")) {
            return null;
        }
        Map<String, Integer> userReports = new HashMap<>();
        for (JsonNode userReport : data.get("user_reports")) {
            userReports.put(userReport.get(0).asText(), userReport.get(1).asInt());
        }

        return userReports;
    }

    /** Gets a map of reasons to the moderator that used that reason to report this submission */
    @JsonProperty(nullable = true)
    public Map<String, String> getModeratorReports() {
        if (!data.has("mod_reports")) {
            return null;
        }
        Map<String, String> modReports = new HashMap<>();
        for (JsonNode userReport : data.get("mod_reports")) {
            modReports.put(userReport.get(0).asText(), userReport.get(1).asText());
        }

        return modReports;
    }


    @Override
    public DistinguishedStatus getDistinguishedStatus() {
        return _getDistinguishedStatus();
    }

    @Override
    public Integer getTimesGilded() {
        return _getTimesGilded();
    }

    @Override
    public Integer getScore() {
        return _getScore();
    }

    @Override
    public VoteDirection getVote() {
        return _getVote();
    }

    /**
     * Checks if this contribution has been archived. If so, then it is no longer open for voting.
     */
    @JsonProperty
    public boolean isArchived() {
        return data("archived", Boolean.class);
    }
}

File: lib/src/main/kotlin/net/dean/jraw/models/PublicContribution.kt
package net.dean.jraw.models

import net.dean.jraw.references.PublicContributionReference
import net.dean.jraw.references.Referenceable
import java.util.*

interface PublicContribution<out T: PublicContributionReference> :
    Created, Distinguishable, Gildable, Identifiable, Votable, Referenceable<T> {

    /** Username of the user that created this model */
    val author: String

    /** The body of the post or comment. Null if the submission is not a self post */
    val body: String?

    /** The date at which the Submission was last edited, or null if it hasn't been. */
    val edited: Date?

    /** If the user has saved this model */
    val isSaved: Boolean

    /** Stickied submissions appear as the the very top when browsing a subreddit by 'hot' */
    val isStickied: Boolean

    /** If reddit is masking this Submission's score */
    val isScoreHidden: Boolean

    /** The name of the subreddit (e.g. "pics") */
    val subreddit: String

    /** The subreddit's full name */
    val subredditFullName: String
}
